packages/react/src/components/Checkbox/CheckboxBase.tsx

```diff
@@ -0,0 +1,80 @@
+import * as React from 'react'
+
+interface CheckboxBaseProps {
+  checked?: boolean
+  onChange?: React.ChangeEventHandler
+  onClick?: React.MouseEventHandler
+  slots?: {
+    root?: any
+    input?: any
+  }
+  slotProps?: {
+    root?: any
+    input?: any
+  }
+  classes?: any
+}
+
+const Testhelper = (prop?: Function, propName?: string) => {
+  if (!prop) {
+    return {}
+  }
+  return {
+    [propName]: prop,
+  }
+}
+
+const CheckboxBase: React.FunctionComponent<CheckboxBaseProps> = props => {
+  const { checked, onChange, onClick, classes = {}, slots = {}, slotProps = {}, ...rest } = props
+  const { root: RootSlot = 'div', input: InputSlot = 'input' } = slots
+  const { root: rootClass, input: inputClass } = classes
+  const { root: rootProps, input: inputProps } = slotProps
+
+  const [isChecked, setIsChecked] = React.useState(!!checked)
+  const realIsChecked = checked === undefined ? isChecked : !!checked
+
+  const onChangeHandler = React.useCallback(
+    (ev: any) => {
+      if (onClick) {
+        onClick(ev)
+      }
+      if (!ev.defaultPrevented) {
+        if (onChange) {
+          onChange(ev)
+        }
+      }
+      if (!ev.defaultPrevented) {
+        setIsChecked(!realIsChecked)
+      }
+    },
+    [setIsChecked, realIsChecked, onChange],
+  )
+
+  const onKeyDown = React.useCallback(
+    (ev: React.KeyboardEvent) => {
+      console.log(ev.keyCode)
+      switch (ev.keyCode) {
+        case 13:
+        case 32: {
+          setIsChecked(!realIsChecked)
```

Comment:
In both Semantic UI React and Stardust there is the concept of [auto controlled state](https://react.semantic-ui.com/#auto-controlled-state) which allows the consumer to take over and control the state from outside by using a combination of the chnge handler and override state prop. It is especially useful in components like Popup or Dialog where the consumer can add  control if the surface should be opened/closed initially or based on external event, as well as additional logic to extend the internal handlers. Maybe this is a simplified version of the API, but it would be good to keep it as it turned out to be very effective and consistent way of controlling the state from outside.

Reply:
Small note, the concept of "auto controlled state" comes from React itself.  Form controls, like the `<input />`, will automatically control their value when the user inputs values.  However, if the dev controls the `value` prop, then it will defer to the dev's props.  If the dev removes the prop, then React resumes control of the `<input />`.

The principles that follow from this is that:
1. Components should work without requiring the consumer to wire them up.  Dropdowns should open on click, inputs should update their value on input, etc.
2. Components should defer to controlled mode if the dev provides props
3. Components should resume control if the developer "unsets" the prop value (i.e. `undefined`).